### PR TITLE
feature: Add IR for noise simulator

### PR DESCRIPTION
**Pull request by yitinche**
Description of changes: 
- Add instructions, result types and share models that are related to noise simulation in IR.

Build directory: [build_files.tar.gz](https://github.com/aws/amazon-braket-schemas-python/files/5104566/build_files.tar.gz)

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [ x ] I have read the [CONTRIBUTING](https://github.com/aws/amazon-braket-schemas-python/blob/main/CONTRIBUTING.md) doc
- [ x ] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/amazon-braket-schemas-python/blob/main/CONTRIBUTING.md#commit-your-change)
- [ x ] I have updated any necessary documentation, including [READMEs](https://github.com/aws/amazon-braket-schemas-python/blob/main/README.md) and [API docs](https://github.com/aws/amazon-braket-schemas-python/blob/main/CONTRIBUTING.md#documentation-guidelines) (if appropriate)

#### Tests

- [ x ] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ x ] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/braket/ir/jaqcd/__init__.py [hunk 1]
```diff
@@ -18,16 +18,21 @@
     XY,
     YY,
     ZZ,
+    Amplitude_Damping,
```
Comment (licedric) [inline src/braket/ir/jaqcd/__init__.py:21]: You should probably use the same naming convention as with the other gates (e.g. AmplitudeDamping instead of Amplitude_Damping).

File: src/braket/ir/jaqcd/program_v1.py [hunk 2]
```diff
@@ -106,9 +116,9 @@ class Program(BraketSchemaBase):
     Attributes:
         braketSchemaHeader (BraketSchemaHeader): Schema header. Users do not need
             to set this value. Only default is allowed.
-        instructions (List[GateInstructions]): List of instructions.
-        basis_rotation_instructions (List[GateInstructions]): List of instructions for
-            rotation to desired measurement bases. Default is None.
+        instructions (List[Union[GateInstructions, NoiseInstructions]]): List of instructions.
+        basis_rotation_instructions (List[Union[GateInstructions, NoiseInstructions]]): List of
```
Comment (licedric) [inline src/braket/ir/jaqcd/program_v1.py:120]: Why should NoiseInstructions be allowed here?

File: src/braket/ir/jaqcd/shared_models.py [hunk 3]
```diff
@@ -129,6 +129,21 @@ class Angle(BaseModel):
     angle: confloat(gt=float("-inf"), lt=float("inf"))
 
 
+class Probability(BaseModel):
+    """
+    Probability for noise channel (floating point).
+
+    Attributes:
+        prob (float): The probability for noise channel.
```
Comment (avawang1) [inline src/braket/ir/jaqcd/shared_models.py:137]: Can we change this to "probability"?

File: src/braket/ir/jaqcd/shared_models.py [hunk 4]
```diff
@@ -129,6 +129,21 @@ class Angle(BaseModel):
     angle: confloat(gt=float("-inf"), lt=float("inf"))
 
 
+class Probability(BaseModel):
```
Comment (avawang1) [inline src/braket/ir/jaqcd/shared_models.py:132]: Can we change the class name to "SingleProbability" because this only accepts 1 float?

File: src/braket/ir/jaqcd/program_v1.py [hunk 5]
```diff
@@ -153,10 +163,17 @@ class Program(BraketSchemaBase):
...
+        Kraus,
     """
 
     _PROGRAM_HEADER = BraketSchemaHeader(name="braket.ir.jaqcd.program", version="1")
     braketSchemaHeader: BraketSchemaHeader = Field(default=_PROGRAM_HEADER, const=_PROGRAM_HEADER)
-    instructions: List[GateInstructions]
+    instructions: List[Union[GateInstructions, NoiseInstructions]]
     results: Optional[List[Results]]
-    basis_rotation_instructions: Optional[List[GateInstructions]]
+    basis_rotation_instructions: Optional[List[Union[GateInstructions, NoiseInstructions]]]
```
Comment (avawang1) [inline src/braket/ir/jaqcd/program_v1.py:179]: +1 same comment as @licedric basis_rotation_instructions are automatically added by the SDK, and NoiseInstructions can't be added to them.

File: src/braket/ir/jaqcd/instructions.py [hunk 6]
```diff
@@ -696,3 +698,104 @@ class Type(str, Enum):
...
+    """
+    Arbitrary quantum channel defined by the input matrices.
+
+    Attributes:
+        type (str): The instruction type. default = "kraus". (type) is optional.
+            This should be unique among all instruction types.
+        targets (List[int]): The target qubits. This is a list with ints and all ints >= 0.
+        matrices (List[List[List[List[float]]]]): A list of matrices specifying
+            the quantum channel. A complex number is represented as a list of 2
+            elemtents. So each matrix has type List[List[List[float]]].
```
Comment (licedric) [inline src/braket/ir/jaqcd/instructions.py:789]: elements (but I'd just write "real numbers" instead)

File: src/braket/ir/jaqcd/shared_models.py [hunk 7]
```diff
@@ -152,6 +167,35 @@ class TwoDimensionalMatrix(BaseModel):
     )
 
 
+class TwoDimensionalMatrixList(BaseModel):
+    """
+    List of Two dimensional non-empty matrices.
```
Comment (licedric) [inline src/braket/ir/jaqcd/shared_models.py:172]: Two shouldn't be capitalized.

Also, it should be "two-dimensional". (Please fix below in comments for matrix as well.)

File: src/braket/ir/jaqcd/program_v1.py [hunk 8]
```diff
@@ -106,9 +117,9 @@ class Program(BraketSchemaBase):
     Attributes:
         braketSchemaHeader (BraketSchemaHeader): Schema header. Users do not need
             to set this value. Only default is allowed.
-        instructions (List[GateInstructions]): List of instructions.
-        basis_rotation_instructions (List[GateInstructions]): List of instructions for
-            rotation to desired measurement bases. Default is None.
+        instructions (List[Union[GateInstructions, NoiseInstructions]]): List of instructions.
+        basis_rotation_instructions (List[GateInstructions]): List of
```
Comment (licedric) [inline src/braket/ir/jaqcd/program_v1.py:121]: Nit: you can revert the changes in basis_rotation_instructions so it doesn't show up in the change history.

File: test/unit_tests/braket/ir/jaqcd/test_probability.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    SingleProbability(probability=float("-inf"))
+
+
+@pytest.mark.xfail(raises=ValidationError)
+def test_negative_float():
+    SingleProbability(probability=float(-1.5))
+
+
+@pytest.mark.xfail(raises=ValidationError)
+def test_great_than_one_float():
```
Comment (licedric) [inline test/unit_tests/braket/ir/jaqcd/test_probability.py:51]: Nit: greater

File: src/braket/ir/jaqcd/program_v1.py (hunk 8 above)
Comment (yitinche) [reply src/braket/ir/jaqcd/program_v1.py:121]: The original changes came with other changes in the same file in the same commit. I don't find a good way to do it. Please let me know if there is a good way to do it, thanks!

**Comment by speller26** [comment]
Completed in #78
